libbeat/esclient/client.go

```diff
@@ -0,0 +1,152 @@
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package client
+
+import (
+	"encoding/json"
+	"errors"
+
+	"github.com/elastic/beats/libbeat/common"
+
+	"github.com/elastic/go-elasticsearch"
+	es6 "github.com/elastic/go-elasticsearch/v6"
+	es7 "github.com/elastic/go-elasticsearch/v7"
+	es8 "github.com/elastic/go-elasticsearch/v8"
+)
+
+// Client provides a version-agnostic Elasticsearch client. It wraps various
+// lower-level version-aware Elasticsearch clients, instantiating the right one
+// for the Elasticsearch cluster being connected to.
+type Client struct {
+	e6 *es6.Client
+	e7 *es7.Client
+	e8 *es8.Client
+
+	version *common.Version
+
+	addr []string
+
+	username string
+	password string
+	apiToken string
+}
+
+type Option func(c *Client) error
+
+// WithAddresses allows you to specify addresses for the Elasticsearch
+// nodes to connect to. Addresses are of the form "http://localhost:9200".
+func WithAddresses(addr ...string) Option {
+	return func(c *Client) error {
+		c.addr = addr
+		return nil
+	}
+}
+
+// WithBasicAuth allows you to specify a username and password to use
+// when authenticating with the Elasticsearch cluster.
+func WithBasicAuth(username, password string) Option {
+	return func(c *Client) error {
+		c.username = username
+		c.password = password
+		return nil
+	}
+}
+
+// WithAPITopken allows you to specify an API token to use
+// when authenticating with the Elasticsearch cluster.
+func WithAPIToken(apiToken string) Option {
+	return func(c *Client) error {
+		c.apiToken = apiToken
+		return nil
+	}
+}
+
+// Constructor
+func New(opts ...Option) (*Client, error) {
+	c := &Client{
+		addr: []string{"http://localhost:9200"},
+	}
+
+	for _, opt := range opts {
+		err := opt(c)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	e, err := elasticsearch.NewClient(elasticsearch.Config{
+		Addresses: c.addr,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	version, err := getClusterVersion(e)
+	if err != nil {
+		return nil, err
+	}
+
+	c.version = version
+
+	switch c.version.Major {
+	case 6:
+		c.e6, err = es6.NewClient(es6.Config{
+			Addresses: c.addr,
+		})
+	case 7:
+		c.e7, err = es7.NewClient(es7.Config{
+			Addresses: c.addr,
+		})
+	case 8:
+		c.e8, err = es8.NewClient(es8.Config{
+			Addresses: c.addr,
+		})
+	default:
+		return nil, ErrUnsupportedVersion
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
```

Comment:
One question I have is when a version _renogation_ is called again? When upgrading your stack the upgrades will be started from ES and the beats after. When the _es client version gets updated_, or does it get updated?

Because this would be static, I like static stuff.

I wish that Elasticsearch had a similar mechanism as Kibana where you send the `kbn-version` and validation happend on the remote side.

For us, maybe we need to periodically check that version or we trigger the verification on some type of errors?